quickwit-config/src/config.rs

```diff
@@ -224,19 +212,56 @@ pub struct StorageConfig {
 }
 
 #[derive(Debug, Serialize, Deserialize, PartialEq)]
-pub struct ServerConfig {
+pub struct QuickwitConfig {
     pub version: usize,
+    #[serde(default = "default_node_id")]
+    pub node_id: String,
+    #[serde(default = "default_metastore_uri")]
     pub metastore_uri: String,
+    #[serde(default = "default_data_dir_path")]
+    pub data_dir_path: PathBuf,
     #[serde(rename = "indexer")]
-    pub indexer_config: Option<IndexerConfig>,
+    #[serde(default)]
+    pub indexer_config: IndexerConfig,
     #[serde(rename = "searcher")]
-    pub searcher_config: Option<SearcherConfig>,
+    #[serde(default)]
+    pub searcher_config: SearcherConfig,
     #[serde(rename = "storage")]
     pub storage_config: Option<StorageConfig>,
 }
 
-impl ServerConfig {
-    pub async fn from_file(path: &str) -> anyhow::Result<Self> {
+pub fn default_quickwit_index_uri(quickwit_config: &QuickwitConfig, index_id: String) -> String {
+    format!(
+        "file://{}",
+        quickwit_config
+            .data_dir_path
+            .join("indexes")
+            .join(index_id)
+            .to_string_lossy()
+    )
+}
+
+fn default_quickwit_config_path() -> io::Result<String> {
+    env::current_dir().map(|path| path.join("config.yaml").to_string_lossy().to_string())
+}
+
+impl QuickwitConfig {
+    // Loads quickwit config from the given path or from the default config file
+    // and validates.
+    pub async fn load(path: Option<String>) -> anyhow::Result<Self> {
+        let config_path = if let Some(config_path) = path {
+            config_path
+        } else {
+            let config_path = default_quickwit_config_path()?;
+            info!("Use default config file: {}", config_path);
```

Comment:
Do we display a good error message when we fail down the line if that file does not exist? I wonder if we should fail ealier.

Reply:
Good point, we can improve a little, will have a look.

Reply:
This issue should be treated separately. The file metastore will create the necessary directories if it has the permissions to do it. To avoid getting an error super late, I modified the "check" function in the file backed metastore, it now creates the directories to check permissions. It's "borderline" but I found it acceptable.

Reply:
Done.